contracts/Staking.sol

```diff
@@ -0,0 +1,488 @@
+  uint8 public constant EVENT_TRANSFER_IN_REWARD = 0x04;
+  uint8 public constant EVENT_TRANSFER_IN_UNDELEGATED = 0x05;
+
+  uint32 public constant ERROR_UNKNOWN_PACKAGE_TYPE = 101;
+  uint32 public constant ERROR_WITHDRAW_BNB = 102;
+
+  uint256 constant public TEN_DECIMALS = 1e10;
+
+  uint256 public constant INIT_ORACLE_RELAYER_FEE = 2e16; //TODO
+  uint256 public constant INIT_MIN_DELEGATION_CHANGE = 1e19;
+  uint256 public constant INIT_CALLBACK_GAS_LIMIT = 23000;
+
+  uint256 public oracleRelayerFee;
+  uint256 public minDelegationChange;
+
+  mapping(address => uint256) delegated;
+  mapping(address => mapping(address => uint256)) delegatedOfValidator;
+  mapping(address => uint256) distributedReward;
+  mapping(address => mapping(address => uint256)) pendingUndelegated;
+  mapping(address => uint256) undelegated;
+
+  bool internal locked;
+
+  modifier noReentrant() {
+    require(!locked, "No re-entrancy");
+    locked = true;
+    _;
+    locked = false;
+  }
+
+  modifier tenDecimalPrecision(uint256 amount) {
+    require(msg.value%TEN_DECIMALS==0,
+      "invalid msg value: precision loss in amount conversion");
+    require(amount%TEN_DECIMALS==0,
+      "invalid amount: precision loss in amount conversion");
+    _;
+  }
+
+  modifier initParams() {
+    if (!alreadyInit) {
+      oracleRelayerFee = INIT_ORACLE_RELAYER_FEE;
+      minDelegationChange = INIT_MIN_DELEGATION_CHANGE;
+      alreadyInit = true;
+    }
+    _;
+  }
+
+  /*********************************** Events **********************************/
+  event delegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event undelegateSubmitted(address indexed delegator, address indexed validator, uint256 amount, uint256 oracleRelayerFee);
+  event redelegateSubmitted(address indexed delegator, address indexed validatorSrc, address indexed validatorDst, uint256 amount, uint256 oracleRelayerFee);
+  event rewardReceived(address indexed delegator, uint256 amount);
+  event rewardClaimed(address indexed delegator, uint256 amount);
+  event undelegatedReceived(address indexed delegator, uint256 amount);
+  event undelegatedClaimed(address indexed delegator, uint256 amount);
+  event failedDelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedUndelegate(address indexed delegator, address indexed validator, uint256 amount, uint8 errCode);
+  event failedRedelegate(address indexed delegator, address indexed valSrc, address indexed valDst, uint256 amount, uint8 errCode);
+  event paramChange(string key, bytes value);
+  event failedSynPackage(uint256 errCode);
+  event crashResponse();
+
+  receive() external payable {}
+
+  /************************* Implement cross chain app *************************/
+  function handleSynPackage(uint8, bytes calldata msgBytes) external onlyCrossChainContract onlyInit override returns(bytes memory) {
```

Comment:
shall we change `onlyInit` to `initParams` here?

Reply:
yes, done